### ISSUE TITLE
Rancher picks a random port once instead of per-container when creating a new service without a fixed port mapping

### ISSUE THREAD

**Issue by iangcarroll** [rancher/rancher#4178]
Rancher Version: v0.63.1

Docker Version: 1.10.3

OS: Ubuntu

When provisioning a stack from `rancher-compose`, Rancher appears to pick one port for a port mapping that isn't fixed (i.e. docker dynamically assigns it) for all containers, which poses a problem when you have more containers of that type than host. 

For example, if I provision the service `web` and specify 80 under `ports` in `docker-compose.yml`, Rancher picks a random port like `53563` and then tries to use it for all containers. If I have three hosts and request eight `web` containers, it becomes stuck. The Rancher UI shows it's only trying to use the port it randomly picked as well.

**Comment by alena1108** [comment]
@deniseschannon  @iangcarroll this is by design. Port defined on the service, should be the same for all containers of the service; whether it's assigned, or picked randomly. As for the case with statically assigned public port, you would need to make sure you have enough hosts in the system to place all service's containers w/o running into port constraints violations

**Comment by iangcarroll** [comment]
I think I poorly worded this issue. You can't fix the issue where a container requests `80:80` and there aren't enough hosts, but when there are more containers than hosts for a service requesting any port mapped to `80` on the container, Rancher shouldn't be running into issues. It already does this when you scale a service up after creating it, so it would probably make sense to do it here too.

**Comment by alena1108** [comment]
> You can't fix the issue where a container requests 80:80 and there aren't enough hosts, but when there are more containers than hosts for a service requesting any port mapped to 80 on the container, Rancher shouldn't be running into issues. It already does this when you scale a service up after creating it, so it would probably make sense to do it here too.

scaling up service when there are not enough hosts to satisfy port requirement, shouldn't be treated any different from creating a service when there are not enough hosts for deployment. In both cases service should be left in intermediate state, and reason for the deployment failure is presented for the user in the UI, so the user can fix it either by a) adding more hosts, or b) scale down the service so scale = number of eligible hosts. 

Today b) - updating the service - is available only for the services that are being scaled up, but not for the services that are being created (activated). @sangeethah @deniseschannon I believe we have a ticket asking to add a support for it, could you link it here?

**Comment by vincent99** [comment]
@alena1108 C) pick another random port available on all hosts and use that one until there are no hosts with that available with that port & repeat.  Trying to be friendly to DNS and use the same port shouldn't prevent you from having scale > hosts with a random port.

**Comment by alena1108** [comment]
@vincent99 that might work only for initial service creation as we can't assign more than one public port to the service's private port, random or not random. So in order to reapply newly picked port to all the instances, we would have to recreate existing ones

**Comment by vincent99** [comment]
Sounds like we need to fix that then :smile:  Need to be able to have random ports with #containers > #hosts, and be able to scale it up & down.  So it will need to use more than 1, and should group them into approximately the minimum number of distinct host ports needed.

**Comment by bengerman13** [comment]
>  So in order to reapply newly picked port to all the instances, we would have to recreate existing ones

This doesn't sound like what (I think) vincent99 is asking for. I think what he's looking for (and definitely what i'm looking for) is something like this:
1. define service with random port mapping
2. rancher picks some random, available port (say 56789)
3. scale service to # containers == # hosts
4. scale service to have one more host
5. Rancher picks a second, different port (sat 56790). Existing containers have no change, new containers get port 56790.

**Comment by alena1108** [comment]
@bengerman13 I understand that. But with the current design, the random port that we chose for a specific private port of the service, is shared by all the containers of the service (=by all the hosts).

@vincent99 @ibuildthecloud I don't recall the exact reasons/requirements for this particular way of implementation over doing it on per individual container basis, but I do remember that was the initial requirement - random port being consistent cross service's container. If you feel that this way is not quite correct, we can change it.

**Comment by vincent99** [comment]
It was just a lot simpler this way.  To do multiple ports will need basically a port scheduler and somehow balancing container scheduling desires vs port availability.  We'll talk offline :smile:

**Comment by alena1108** [comment]
@vincent99 ok

**Comment by demohi** [comment]
@alena1108  Why rancher expose the port for developer? I think if  they need it ,they can define it in docker-compose file or use load balancer to expose the same port in multi hosts

With this design, how can I run multi container(same image) in a host ?

**Comment by iangcarroll** [comment]
Rancher will handle `expose` correctly, just not `ports`. In most cases this is fine because you can put them behind a load balancer.

**Comment by akurkin** [comment]
This issue been bugging me for a long time, but after spending a bit more time thinking and playing around with rancher and different types of services, I came up with conclusion that current design makes the most sense:
- if as part of docker-compose file you've specified that your service is available on port XX, you want to have consistent "entry points" for that service across all machines
- if your dockerized app doesn't really care at which public port it runs and only concerned about internal private port (in docker container), then just use `EXPOSE` instruction in your Dockerfile and it will work beautifully allowing you to scale app to multiple instances even on a single machine

What would be really cool, is to **put these guidelines / recommendations somewhere in the docs**. Because in case of providing only private port in docker-compose, docs don't say anywhere that port will be randomly chosen only once (not for every container). And, you know, you have an expectation that it will be chosen randomly for every container (if you replicate same scenario via vanilla docker-compose features).

**Comment by remkade** [comment]
@akurkin are you saying that rancher currently works with `expose`? I can't get the expose directive in docker compose to be respected and we don't have a ui button for it.

**Comment by akurkin** [comment]
@remkade I mentioned `EXPOSE` instruction in Dockerfile. I haven't used expose in docker-compose.yml. So basically Dockerfile keeps the knowledge about private ports vs docker-compose

**Comment by remkade** [comment]
@akurkin I'm not sure how to use that in Rancher. Does rancher support some way of doing the docker `--publish-all`?

**Comment by akurkin** [comment]
@remkade sorry I have no idea. I'm using it like this:

Assume I have service `api`, then:
- set `EXPOSE 3000` in `Dockerfile` (for conventions I used port 3000 for every service)
- do not set any port mappings in `docker-compose.yml`

From load balancer, or any other service, I can do `curl http://api:3000/` -- notice I didn't publish any ports via docker-compose, yet it's still available when querying directly my private port (thanks to rancher networking)

**Comment by dishcandanty** [comment]
@vincent99 @alena1108 any update on this? (Picking a new random port, rather than a random port when service is created)

This would be extremely nice :)

**Comment by jhmartin** [comment]
This also makes sense when using Consul Registrator, since you are then fine w/having dynamic port numbers.

**Comment by InAnimaTe** [comment]
^exactly.

There are a couple extra points I'd like to make regarding port usability scenarios that might be worth offering at some point in the future:

1. Since a multitude of people already use Service Discovery methods, in addition to utilizing rancher-metadata for whatever service catalog people might need, some users may want the option to enable random port assignments for every deployment of the container. One reason for this would be to cease from exhausting a specific port on the cluster (applicable to SaaS use-cases). With that, since that port isn't actually reserved on hosts which don't yet have the service running, when trying to expand scale in the cluster, that same port may not be obtainable (another service may be using it).

2. Presently, this functionality only seems to work for Service deployed containers, not for just flat containers which are launched (and have the same port mapping configuration available). To me, it seems pretty trivial to provide this for launched one-off containers, whereas that individual container gets a random port assigned to it which isn't reserved (as it isn't a service). There was some talk about this in [irc](https://botbot.me/freenode/rancher/msg/73323473/).

**Comment by uberamd** [comment]
+1 to this feature request. Similar to what @jhmartin said, I'm using Registrator with Consul and Consul-Template to add services to automatically created load balancers that live on their own network (not containers).

Ideally I'd be able to have a random host port mapping to, say, port 3000 in my container and scale up > the number of hosts, with a random port being assigned to each scaled container. Registrator + Consul handles this fine and updates my external load balancer accordingly.

The workaround being used now is to `EXPOSE` the container port but create no hosts ports, then create a load balancer service within Rancher that has an appropriate `SERVICE_NAME` label defined. The downfall to this is that the traffic is hopping from my external load balancer -> Rancher load balancer -> container vs external load balancer -> container.

**Comment by ronpanto** [comment]
+1
I am looking into rancher now only because i thought it would be able to do that out of the box. AWS ECS is able to do that for http+https and i need regular TCP.
There is a huge difference between running out of ports in the cluster to other resources since usually we are talking about a different order of magnitude.

**Comment by n0rig** [comment]
I am actually running multiple services on ONE server and need the Docker built in port forwarding to handle the port assigning. Otherwise, when I try to scale my services they all collide with the first. At least make it possible to support NOT assigning the same ports for ALL the service containers.

If I run them from the command line individually, everything works perfectly. But, when I try to use the rancher/cattle environment, all of the service containers collide with the first one because the dynamic port assignment, which docker performs, is not being carrier out since all the new containers use the first container's port assignment.

#7400

**Comment by chrisjenx** [comment]
I've noticed this on 1.4.1.

- Create a service with portmap: public port empty and private container port set.
- Create Service
![screen shot 2017-03-03 at 17 02 38](https://cloud.githubusercontent.com/assets/1167793/23560825/a1c7bbd2-0033-11e7-90a0-62eb75a0947b.png)
-It populates the pub port once instead of randomly at runtime.

However, if I create the container first with NO port mappings. Run it, Upgrade it by adding a port mapping, then it leaves the pub port blank and works as intented of randomly picking a port per container.

**Comment by chrisjenx** [comment]
After reading the comments above - this is bad design by Rancher.

- If you define EXPOSE on a docker file the host should choose a random public port for that container.
- Rancher will statically assign a public port for that container. That is inconsistant with docker thus this entire chain.

I would either:
1. Change the wording to make it clearer that this is public port is assigned across all the services.
2. Change the functionality (docker/aws/kubernetes etc) blank = random port.

2. would actually match the docker compose spec and reduce this confusion.

**Comment by Moeser** [comment]
The current behavior does not match the behavior of docker or docker-compose, and creates another pain point for people migrating from a non-rancher orchestration style.  @vincent99 @alena1108 the last update from Rancher on this issue is from over a year ago, has there been any work on this, or are there any updates you can provide?

On a more technical note: Is there any reason this behavior is considered better than how docker-compose would handle port assignment?  At the very least, can we get an option to retain the same behavior as docker/docker-compose?  Adding an entire port scheduling system in Rancher seems overkill, when docker is already able to schedule the port on a per-host basis.  Couldn't you just allow docker to schedule the port, and have Rancher read the port that docker scheduled?  At that point, showing the scheduled port correctly in the Rancher API and rancher-metadata would be enough to unblock us.

**Comment by vincent99** [comment]
I have nothing new to add really, and clearly there hasn't been work on it.

- `EXPOSE` does not publish ports to the host (in Rancher, regular docker, or anywhere else).  It just declares that the container listens on those ports.

- The behavior of docker and docker-compose is irrelevant because it only deals with one container/host.  Things like external-dns are useless if the port being exposed varies with every host IP.  So you then need some other non-standard mechanism to discover all the ip+port combinations that are published and, clients that support that mechanism.

- The docker daemon is not involved in picking port numbers, and in fact doesn't even know ports are published at all in modern Rancher versions, because networking and port mapping are handled by CNI as iptables rules and the actual container is `--net=none`.

- Swarmkit services implements a similar behavior, except that it reserves the port on *every* host, whether it has containers running for that service or not.

**Comment by robertpearce** [comment]
that makes it sound like exposing a port with ports: 80 in a docker compose file of a stack should allow it to scale where scale > hosts ? That's not the behavior i'm seeing. What i'm seeing is the stack getting converted to be ports: 12345:80 (say) and then additional copies of the service clashing on port usage.

the problem i have is that i have scale > hosts, but the scaled nodes are not identical. I'm running behat testing against many workers and while each worker node is the same docker image, the files on disk differ after boot time, hence while i can (and do) load balance incoming requests via an haproxy container to give a shared endpoint for testing against and then use a service selector to find the worker node to send the request to, the current rancher inability for scale > hosts on the worker nodes kills my use case, as far as i can tell.

I need to be able to have more than 1 copy of a service on hosts, while exposing a port (any port, i don't care which) for a load balancer service that lives on anouther host to access. It sounds like this bug is a blocker for my use case currently as rancher seems to be assigning a local port for that service, and then choking when i go to scale it up.

**Comment by Moeser** [comment]
Referring to @vincent99 's message further above:

> Things like external-dns are useless if the port being exposed varies with every host IP.

If you assume all applications only ever use a single fixed port, and never use any service discovery, then yes you might run into problems using external dns.  Anything that uses consul or zookeeper or another service discovery mechanism can also register their port.  I feel like service discovery is a common practice in the docker world, and a blind spot for Rancher when it comes to issues like this.  Assuming a service will only ever use a single fixed port is applicable in the pre-docker world, where you probably only ever run one copy of the service per-host.  In a dockerized world, does it really make sense to keep following that assumption?

> So you then need some other non-standard mechanism to discover all the ip+port combinations that are published, and clients that support that mechanism.

I wouldn't call consul and zookeeper non-standard.  And having clients that support them is pretty common.

> The behavior of docker and docker-compose is irrelevant because it only deals with one container/host.

I get what you're saying here, but you could use that argument for almost anything native-docker vs rancher.  Just because docker was originally designed with individual hosts in mind, doesn't mean they made bad decisions for multi-host environments, remember that they've been working on docker-swarm for a while.  The point is that the direction docker went with (randomized external ports if one isn't specified) *would* fix the scale > num_hosts problem, and forcing every service to have a single port will always re-introduce the scaling problem.  Some of us picked rancher because we *don't* want to treat our hosts like pets.  If I can only run one copy of a container on a host, even though the host is totally underutilized otherwise, then I have to manually balance something else onto it to make full use of those resources, and that's getting further away from treating our hosts like cattle.

Come on guys, less pets, more cattle.

**Comment by vincent99** [comment]
Again, we're aware of the issue and have nothing to add today.  And there is a perfectly serviceable workaround for most people; don't publish a port, run as many as you want, and make them available to the internet using a limited set of `n < hosts` load balancer containers.

**Comment by Moeser** [comment]
> don't publish a port, run as many as you want, and make them available to the internet using a limited set of n < hosts load balancer containers.

That workaround only works if your service is capable of running behind rancher's load balancer.  Unfortunately not all services fit into that mold.  The situation gets a lot more problematic with UDP based services, kerberos or mutually auth'd services, and services which need to accurately advertise their address (and probably several other cases that we haven't even run into yet).  

The reason this issue is important is not because of the 90% of services that work fine behind a load balancer, it's important because of the rest which do not.  We'd do things the way you're suggesting if that was an option, but in a lot of cases it's not.  If the load balancer worked for all applications, this wouldn't even be an issue.

There's a tendency to view any non-http based service as if we're trying to do some wacky off-the-wall thing in some unsupported way.  I don't feel like it's the fault of the users that they're running an application which doesn't work behind the rancher LB and need it to scale.

**Comment by McSnowflake** [comment]
I am running a selenium grid on docker. All the nodes have the same internal port, thats why I need Rancher to expose them on different external ports. They will then tell the hub to use this externel port to communicate. Currently I can only run such a service with scale one.

**Comment by dvmedvedev** [comment]
@McSnowflake
I'm using this schema in Rancher and had no problem.
```
version: '2'
services:
  node-chrome:
    image: selenium/node-chrome:3.0.1-carbon
    environment:
      HUB_PORT_4444_TCP_ADDR: selenium-hub
      HUB_PORT_4444_TCP_PORT: '4444'
    links:
      - selenium-hub:hub
  node-firefox:
    image: selenium/node-firefox:3.0.1-carbon
    environment:
      HUB_PORT_4444_TCP_ADDR: selenium-hub
      HUB_PORT_4444_TCP_PORT: '4444'
    links:
      - selenium-hub:hub
  selenium-hub:
    image: selenium/hub:3.0.1-carbon
    ports:
      - 4444:4444/tcp
      - 57086:80/tcp
```

**Comment by warroyo** [comment]
Has there been any movement on this? 

I would also like to be able to have the random port selection work on a per container basis. I am using an external LB(F5) and the container get automatically registered to the F5 pools so the ports do not need to be the same since this is dynamic. being able to run multiple instances of an application without having the same number for hosts seems very useful.

**Comment by vincent99** [comment]
This is not ever going to change for 1.x at this point.

For 2.0 nodePorts work in a similar way, a single port is allocated.  But it's allocated/used on _all_ nodes, independent of how many pods there are or where they are running, and then that port on any node routes to however many pods there are.

**Comment by warroyo** [comment]
@vincent99 thanks for the info. what is the reason it is able to choose different ports per container when upgrading and adding a port mapping?

**Comment by peshay** [comment]
I think to have random external ports, for each service an own, would be really good for my current use case:
I have an application that connects to a server and tells the server how it can connect back to the application.
In my container I made scripts to get the external port and external IP from rancher-metadata in order to add it to the application configuration. Just after I made this up, I recognized that I just can scale to the number of allowed hosts, which in my case is 2 :(
Does anyone has an idea how else I could solve that? Only thing I see now is to make the docker-compose fix on every service I want to have and not use dynamic scaling :(

Btw when I tell the application a port, it will also start on that port which is the external port and not the internal one. My workaround for that to do redirection in that container via ncat. So portwise it goes like this:
56431 -> 9099 -> 56431
